test/test/DottyDocTests.scala

```diff
@@ -0,0 +1,506 @@
+package test
+
+import dotty.tools.dotc.Compiler
+import dotty.tools.dotc.core.Phases.Phase
+import dotty.tools.dotc.typer.FrontEnd
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.ast.Trees._
+
+/** All tests need to extend this trait and define `source` and `assertion`
+ *
+ * Instances also need to be placed in the `DottyDocTests::tests` sequence to
+ * be included in the run
+ */
+trait DottyDocTest extends DottyTest { self =>
+  /** Source code in string format */
+  def source: String
+
+  /** A test to run against the resulting code */
+  def assertion: PartialFunction[Tree[Untyped], Unit]
```

Comment:
For any non-trivial check\error reporting, you would need an appropriate context.
Have a look how it's done in `DottyTest` it has a lot of infrastructure that you can use.